src/pyhf/infer/intervals.py

```diff
@@ -15,7 +17,132 @@ def _interp(x, xp, fp):
     return tb.astensor(np.interp(x, xp.tolist(), fp.tolist()))
 
 
-def upperlimit(data, model, scan, level=0.05, return_results=False, **hypotest_kwargs):
+def upperlimit_auto(
+    data,
+    model,
+    low,
+    high,
+    level=0.05,
+    atol=2e-12,
+    rtol=None,
+    calctype='asymptotics',
+    test_stat='qtilde',
+    from_upperlimit_fn=False,
+):
+    """
+    Calculate an upper limit interval ``(0, poi_up)`` for a single
+    Parameter of Interest (POI) using an automatic scan through
+    POI-space, using the :func:`~scipy.optimize.toms748` algorithm.
+
+    Example:
+        >>> import numpy as np
+        >>> import pyhf
+        >>> pyhf.set_backend("numpy")
+        >>> model = pyhf.simplemodels.hepdata_like(
+        ...     signal_data=[12.0, 11.0], bkg_data=[50.0, 52.0], bkg_uncerts=[3.0, 7.0]
+        ... )
+        >>> observations = [51, 48]
+        >>> data = pyhf.tensorlib.astensor(observations + model.config.auxdata)
+        >>> obs_limit, exp_limits = pyhf.infer.intervals.upperlimit_auto(
+        ...     data, model, 0., 5.
+        ... )
+        >>> obs_limit
+        array(1.01156939)
+        >>> exp_limits
+        [array(0.55988001), array(0.75702336), array(1.06234693), array(1.50116923), array(2.05078596)]
+
+    Args:
+        data (:obj:`tensor`): The observed data.
+        model (~pyhf.pdf.Model): The statistical model adhering to the schema ``model.json``.
+        low (:obj:`float`): Lower boundary of search region
+        high (:obj:`float`): Higher boundary of search region
+        level (:obj:`float`): The threshold value to evaluate the interpolated results at.
+                              Defaults to 0.05.
+        atol (:obj:`float`): Absolute tolerance. Defaults to 1e-12. The iteration will end when the
+                             result is within absolute *or* relative tolerance of the true limit.
+        rtol (:obj:`float`): Relative tolerance. For optimal performance this argument should be set
+                             to the highest acceptable relative tolerance, though it will default
+                             to 1e-15 if not set.
+        calctype (:obj:`str`): Calculator to use for hypothesis tests. Choose 'asymptotics' (default)
+                               or 'toybased'.
+        test_stat (:obj:`str`): Test statistic to use. Choose 'qtilde' (default), 'q', or 'q0'.
+
+    Returns:
+        Tuple of Tensors:
+
+            - Tensor: The observed upper limit on the POI.
+            - Tensor: The expected upper limits on the POI.
+    """
+    if rtol is None:
+        _warn(
+            "upperlimit_auto: rtol not provided, defaulting to 1e-15. "
+            "For optimal performance rtol should be set to the highest acceptable relative tolerance."
+        )
+        rtol = 1e-15
+
+    cache = {}
+
+    def f_all(mu):
```

Comment:
this should just use the `lru_cache` in python

Reply:
can't use `lru_cache` here easily. gotta roll our own to access all values that `f_all` has been called with.